examples/gcf-pubsub-vm-delete-event-handler/main.py

```diff
@@ -0,0 +1,660 @@
+from typing import List
+
+
+class RuntimeState:
+    """Stores App instance for the lifetime of the process"""
+    pass
+
+
+RuntimeState.app = None
+
+
+class Result(Enum):
+    """The overall result of the DNS cleanup for user reporting
+
+    An OK result indicates the cleanup completed normally.
+
+    A NOT_PROCESSED results is likely a result of losing the race against the
+    VM delete operation and is intended to signal to the user they may need to
+    cleanup DNS records using another mechanism (e.g. manually).
+    """
+    OK = 0
+    NOT_PROCESSED = 1
+
+
+class Severity(Enum):
+    """Stackdriver severity levels
+
+    See https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry
+    """
+    DEFAULT = 0
+    DEBUG = 100
+    INFO = 200
+    NOTICE = 300
+    WARNING = 400
+    ERROR = 500
+    CRITICAL = 600
+    ALERT = 700
+    EMERGENCY = 800
+
+
+class Detail(Enum):
+    """Detailed results of the cleanup
+
+    A LOST_RACE result indicates user intervention is necessary.
+    """
+    NO_OP = 0
+    NO_MATCHES = 1
+    RR_DELETED = 2
+    VM_NO_IP = 3
+    IGNORED_EVENT = 4
+    LOST_RACE = 5
+    RR_MISMATCH = 6
+    RR_NOT_A_RECORD = 7
+    RR_NAME_MISMATCH = 8
+    RR_IP_MISMATCH = 9
+
+
+class StructuredLog(object):
+    """Base class to report structured log events
+
+    Sub classes are expected to override MESSAGE (str), RESULT (Result), and
+    DETAIL (Detail).
+
+    """
+    MESSAGE = '{} mark'
+    RESULT = None
+    DETAIL = None
+    SEVERITY = Severity.NOTICE
+    LEVEL = logging.INFO
+
+    def __init__(self, vm_uri: str, event_id: str):
+        self.vm_uri = vm_uri
+        self.event_id = event_id
+        self.function_project = os.getenv('GCP_PROJECT')
+        self.function_region = os.getenv('FUNCTION_REGION')
+        self.function_name = os.getenv('FUNCTION_NAME')
+        stream = os.getenv('DNS_VM_GC_REPORTING_LOG_STREAM')
+        # The log stream structured reports are sent to.
+        # https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry
+        if stream:
+            self.log_name = stream
+        else:
+            self.log_name = (
+                'projects/{}/logs/{}'
+            ).format(self.function_project, self.function_name)
+
+    def message(self):
+        """Returns a human readable log message"""
+        return self.MESSAGE.format(self.vm_uri)
+
+    def result(self):
+        """Result code of the DNS cleanup, e.g. OK or NOT_PROCESSED"""
+        return self.RESULT.name
+
+    def detail(self):
+        """Detail code of the DNS cleanup, e.g. NO_OP, RR_DELETED, LOST_RACE"""
+        return self.DETAIL.name
+
+    def severity(self):
+        """Stackdriver Log Severity
+
+        https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry
+        """
+        return self.SEVERITY.name
+
+    def info(self):
+        """Assembles dict intended for jsonPayload"""
+        return {
+            'message': self.message(),
```

Comment:
Will take another look at this, but I'm a bit confused on where the behavior of interpolating the message should move to.  The initializer?

Reply:
see updated comment just do the format right here in `info(self)`

Reply:
Understood thanks.  I'm fixing all of these up now.

I'm going to keep message() as the sole getter because some of the structured logs take more than just the vm_uri.  For example:

```python
class RecordSkipped(StructuredLog):
    MESSAGE = "{} does not match DNS record {} ({})"

    def message(self):
        return self.MESSAGE.format(self.vm_uri, self.record['name'], self.DETAIL.name)
```

Reply:
Fixed in 4a4326bbd7332ebf3e96d764ef60fbdd8965243a